python/pyspark/sql/functions.py

```diff
@@ -2840,6 +2840,463 @@ def from_csv(col, schema, options={}):
     return Column(jc)
 
 
+def _unresolved_named_lambda_variable(*name_parts):
+    """
+    Create `o.a.s.sql.expressions.UnresolvedNamedLambdaVariable`,
+    convert it to o.s.sql.Column and wrap in Python `Column`
+
+    :param name_parts: str
+    """
+    sc = SparkContext._active_spark_context
+    name_parts_seq = _to_seq(sc, name_parts)
+    expressions = sc._jvm.org.apache.spark.sql.catalyst.expressions
+    return Column(
+        sc._jvm.Column(
+            expressions.UnresolvedNamedLambdaVariable(name_parts_seq)
+        )
+    )
+
+
+def _get_lambda_parameters(f):
```

Comment:
@zero323, can we remove `_get_lambda_parameters` and `_get_lambda_parameters_legacy` here too for now? Let's discuss and figure out a better way in the next PR about this.

Reply:
Sorry, but I don't see this.

- Argument type validation is not covered by planner, and some combinations of argument types will either lead to cryptic exceptions or ambiguous behavior. How would you generate placeholders for example for `*args` (or similarly `...` in R).
- Having consistent placeholders with Scala counterpart is good. And even to get to `x1`, x2`, `...,`  cannot be done without inspecting signature. And even if we did only that (which still keeps most of the complexity, calling things `_(**kwargs)` or `_(*args)` or different keyword variants would be rather sloppy.

Reply:
I am not particularly against this codes. I just thought it's easier to logically separate two PRs. This PR just to add higher order functions relying on JVM check. The other PR to add argument checking in Python (or R) sides.

For the former case, I am totally supportive and I can sign-off. For the latter I am not completely sure. I would like to collect some feedback from some more people. So .. asking to remove it is just to make it rolling quicker :-).

Reply:
I get you point but right now I consider this core logic. We can shuffle things around and shave a line or two, but I don't see it improving overall "reviewability".

Somewhere here we have to decide out how many placeholders  should be provided (if it wasn't for polymorphic behavior of some variants, we could just assume ‒ if we for example had `transform` and `transform_with_index`. But now have to do it at least for a few of functions, and there might be more to come). So we can only push things around and

```python
    import inspect

    signature = inspect.signature(f)
    parameters = signature.parameters.values()
```

as well as the legacy part (I hope it is going to be dropped before 3.1 release anyway) are here to stay (could be inlined, but personally I think it would makes things worse not better).

If we wanted to condense things as much as possible, it could be condensed into something like

```python

    args = [
        _unresolved_named_lambda_variable(f"x{i}") for i, _ in 
        enumerate(
          inspect.signature(f).parameters if sys.version_info >= (3, 3) 
          else inspect.getargspec(f) 
        )
    ]

```

but I don't think that it makes things any clearer, do you?  Also leaves the below.


If we keep just this part and leave variadic argument validation out, I'd be concerned that it will to all kinds of confusing or potentially incorrect behaviors. Users are extremely imaginative and delayed resolution doesn't help*. 

And since it is neither heavy in logic nor duplicates (for some definition of duplication) JVM logic, it shouldn't be controversial. 




----------------
\*  We can be pretty sure that  attempts to apply some-general-purpose-library in func will happen, if this is merged. 

And let's be honest ‒ higher order functions are already pretty hard to work with. There are sensitive to minor type mismatches (numeric precision and such) and corresponding expressions are pretty large.

Reply:
Just to be clear ‒ my reservations are now purely practical (in contrast to my opposition to removal of function specific validation). Certain things have to be done one way or another, some parts of the logic will be removed soon so it is better to keep these clearly separated and preventing ambiguity while keeping things || with Scala  is just basic sanity :)